src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs

```diff
@@ -0,0 +1,111 @@
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *-------------------------------------------------------------------------------------------*/
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.Unity.Analyzers.Resources;
+
+namespace Microsoft.Unity.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class ProtectedUnityMessageAnalyzer : DiagnosticAnalyzer
+	{
+		internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+			id: "UNT0021",
+			title: Strings.ProtectedUnityMessageDiagnosticTitle,
+			messageFormat: Strings.ProtectedUnityMessageDiagnosticMessageFormat,
+			category: DiagnosticCategory.Correctness,
+			defaultSeverity: DiagnosticSeverity.Info,
+			isEnabledByDefault: true,
+			description: Strings.ProtectedUnityMessageDiagnosticDescription);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterSyntaxNodeAction(AnalyzeMethodDeclaration, SyntaxKind.MethodDeclaration);
+		}
+
+		private static void AnalyzeMethodDeclaration(SyntaxNodeAnalysisContext context)
+		{
+			var method = (MethodDeclarationSyntax)context.Node;
+			if (method?.Body == null)
+				return;
+
+			if (method.Modifiers.Any(SyntaxKind.ProtectedKeyword))
+				return;
+
+			var classDeclaration = method.FirstAncestorOrSelf<ClassDeclarationSyntax>();
+			if (classDeclaration == null)
+				return;
+
+			var typeSymbol = context.SemanticModel.GetDeclaredSymbol(classDeclaration);
+			var scriptInfo = new ScriptInfo(typeSymbol);
+			if (!scriptInfo.HasMessages)
+				return;
+
+			var symbol = context.SemanticModel.GetDeclaredSymbol(method);
+			if (!scriptInfo.IsMessage(symbol))
+				return;
+
+			context.ReportDiagnostic(Diagnostic.Create(Rule, method.Identifier.GetLocation()));
+		}
+	}
+
+	[ExportCodeFixProvider(LanguageNames.CSharp)]
+	public class ProtectedUnityMessageCodeFix : CodeFixProvider
+	{
+		public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(ProtectedUnityMessageAnalyzer.Rule.Id);
+
+		public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+			var declaration = root.FindNode(context.Span) as MethodDeclarationSyntax;
```

Comment:
root can be null and we can use pattern matching:

```suggestion
				if (!(root?.FindNode(context.Span) is MethodDeclarationSyntax declaration))
				return;
```